pkg/primitives/volume/volume.go

```diff
@@ -0,0 +1,99 @@
+package volume
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+	"github.com/threefoldtech/zos/pkg/gridtypes/zos"
+	"github.com/threefoldtech/zos/pkg/provision"
+	"github.com/threefoldtech/zos/pkg/stubs"
+)
+
+type Volume = zos.Volume
+type VolumeResult = zos.VolumeResult
+
+type Manager struct {
+	client zbus.Client
+}
+
+var (
+	_ provision.Manager = (*Manager)(nil)
+	_ provision.Updater = (*Manager)(nil)
+)
+
+func NewManager(client zbus.Client) Manager {
+	return Manager{client: client}
+}
+
+func (m Manager) Provision(ctx context.Context, wl *gridtypes.WorkloadWithID) (interface{}, error) {
+	storage := stubs.NewStorageModuleStub(m.client)
+
+	var volume Volume
+	if err := json.Unmarshal(wl.Data, &volume); err != nil {
+		return nil, fmt.Errorf("failed to parse workload data as volume: %w", err)
+	}
+	volumeName := wl.ID.String()
+
+	exists, err := storage.VolumeExists(ctx, volumeName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup volume with name %q: %w", volumeName, err)
+	} else if exists {
+		return VolumeResult{ID: volumeName}, provision.ErrNoActionNeeded
+	}
+
+	_, err = storage.VolumeCreate(ctx, volumeName, volume.Size)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new volume with name %q: %w", volumeName, err)
+	}
+	return VolumeResult{ID: volumeName}, nil
```

Comment:
This is of course correct, but I think it's safer to use the Name returned by the VolumeCreate. This ensure things will keep working if we changed the implementation of VolumeCreate for example to return a different name/id of the volume.